### PR TITLE
Add `ValueGraphLoader`

### PR DESCRIPTION
**Pull request by kwridan**
### Short description 📝

@pepibumur has been working on replacing the reference graph (`Graph`) with value based graph (`ValueGraph`). As part of this effort, one of the components needed for this migration is a dedicated loader than can load and convert models to the `ValueGraph` directly.

### Solution 📦

- Add a new dedicated `ValueGraphLoader`

### Implementation 👩‍💻👨‍💻

- [x] Adding a new loader that can load and convert models to a `ValueGraph`
- [x] Update metadata loaders to allow loading metadata needed for external dependencies in one go (equivalent to the graph node loaders)
- [x] Add a new metadata provider for system frameworks (hosts the logic that used to be part of the deprecated SDKNode)

### Test Plan 🛠

- Verify unit tests pass

### Notes  📝

- To ease reviews, this PR only introduces the new component alongside its tests without integrating it, that will be done in a separate PR.
- The previous `GraphLoader` needed a separate component called a `ModelLoader` to load the individual models - this is no longer needed as the generation pipeline has evolved since that was first introduced. All models are now loaded concurrently upfront as such by the time we come to loading the graph we already have all the models available. (see [Generation Pipeline](https://tuist.io/docs/contribution/generation-pipeline/#pipeline) for reference)

### CODE REVIEW

File: Sources/TuistCore/MetadataProviders/FrameworkMetadataProvider.swift [hunk 1]
```diff
@@ -2,7 +2,25 @@ import Foundation
 import TSCBasic
 import TuistSupport
 
+// MARK: - Metadata
+
+public struct FrameworkMetadata: Equatable {
```
Comment (pepibumur) [inline Sources/TuistCore/MetadataProviders/FrameworkMetadataProvider.swift:7]: Nitpick: What if we extract this one to its own file?

File: Sources/TuistCore/MetadataProviders/FrameworkMetadataProvider.swift [hunk 2]
```diff
@@ -18,15 +36,39 @@ public protocol FrameworkMetadataProviding: PrecompiledMetadataProviding {
...
+        guard fileHandler.exists(path) else {
+            throw FrameworkMetadataProviderError.frameworkNotFound(path)
+        }
+        let binaryPath = self.binaryPath(frameworkPath: path)
+        let dsymPath = self.dsymPath(frameworkPath: path)
+        let bcsymbolmapPaths = try self.bcsymbolmapPaths(frameworkPath: path)
+        let linking = try self.linking(binaryPath: binaryPath)
+        let architectures = try self.architectures(binaryPath: binaryPath)
+        let isCarthage = path.pathString.contains("Carthage/Build")
+        return FrameworkMetadata(
```
Comment (pepibumur) [inline Sources/TuistCore/MetadataProviders/FrameworkMetadataProvider.swift:53]: Much better 👏

File: Sources/TuistCore/MetadataProviders/FrameworkMetadataProvider.swift [hunk 3]
```diff
@@ -35,12 +77,38 @@ public final class FrameworkMetadataProvider: PrecompiledMetadataProvider, Frame
...
     }
+
+    private func binaryPath(frameworkPath: AbsolutePath) -> AbsolutePath {
+        frameworkPath.appending(component: frameworkPath.basenameWithoutExt)
+    }
+}
+
+// MARK: - Error
+
+enum FrameworkMetadataProviderError: FatalError, Equatable {
```
Comment (pepibumur) [inline Sources/TuistCore/MetadataProviders/FrameworkMetadataProvider.swift:96]: Nitpick: We've been placing the errors at the top of the file so I'd do the same here to be consistent.

File: Sources/TuistCore/MetadataProviders/LibraryMetadataProvider.swift [hunk 4]
```diff
@@ -2,19 +2,82 @@ import Foundation
...
-final class LibraryMetadataProvider: PrecompiledMetadataProvider, LibraryMetadataProviding {
-    func product(library: LibraryNode) throws -> Product {
-        switch try linking(binaryPath: library.path) {
-        case .dynamic:
-            return .dynamicLibrary
-        case .static:
-            return .staticLibrary
+// MARK: - Error
+
+enum LibraryMetadataProviderError: FatalError, Equatable {
```
Comment (pepibumur) [inline Sources/TuistCore/MetadataProviders/LibraryMetadataProvider.swift:59]: Like I mentioned in another comment, I'd move this one up to the top of the file.

File: Sources/TuistCore/MetadataProviders/SystemFrameworksMetadataProvider.swift [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+// MARK: - Metadata
+
+public struct SystemFrameworkMetadata: Equatable {
+    var name: String
+    var path: AbsolutePath
+    var status: SDKStatus
+    var source: SDKSource
+}
+
+public enum SDKSource: Equatable {
```
Comment (pepibumur) [inline Sources/TuistCore/MetadataProviders/SystemFrameworksMetadataProvider.swift:14]: Nitpick: What about extracting `SDKSource` and `SDKType` into their own files?

File: Sources/TuistCore/MetadataProviders/SystemFrameworksMetadataProvider.swift [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+                    .appending(RelativePath("usr/lib"))
+                    .appending(component: name)
+            }
+        }
+    }
+}
+
+// MARK: - Errors
+
+public enum SystemFrameworkMetadataProviderError: FatalError, Equatable {
```
Comment (pepibumur) [inline Sources/TuistCore/MetadataProviders/SystemFrameworksMetadataProvider.swift:101]: I'd move this one to the top of the fie.

File: Sources/TuistCore/MetadataProviders/XCFrameworkMetadataProvider.swift [hunk 7]
```diff
@@ -73,3 +83,34 @@ class XCFrameworkMetadataProvider: PrecompiledMetadataProvider, XCFrameworkMetad
         return binaryPath
     }
 }
+
+// MARK: - Errors
+
+enum XCFrameworkMetadataProviderError: FatalError, Equatable {
```
Comment (pepibumur) [inline Sources/TuistCore/MetadataProviders/XCFrameworkMetadataProvider.swift:89]: I'd move this one to the top of the file.

File: Sources/TuistCoreTesting/MetadataProviders/MockLibraryMetadataProvider.swift [hunk 8]
```diff
@@ -3,12 +3,37 @@ import TSCBasic
...
+            return LibraryMetadata.test(
+                path: path,
+                publicHeaders: publicHeaders,
+                swiftModuleMap: swiftModuleMap
+            )
         }
     }
 }
+
+public extension LibraryMetadata {
```
Comment (pepibumur) [inline Sources/TuistCoreTesting/MetadataProviders/MockLibraryMetadataProvider.swift:23]: If we end up extracting those into their own files, I'd extract this one into `LibraryMetadata+TestData.swift`.

**Comment by kwridan** [comment]
Thanks for the review @pepibumur - will apply the suggested changes 👍

**Comment by pepibumur** [comment]
@kwridan should we merge this one?

**Comment by kwridan** [comment]
Yes should be good for another review 🙂 - applied the suggestions from the last round, let me know if you spot any others, thanks!

**Review by pepibumur — APPROVED** [review]

**Pull request merged**